luaui/Widgets/cmd_context_build.lua

```diff
@@ -124,25 +176,46 @@ function widget:Update(deltaTime)
 	end
 	local unitDefID = -cmd_id
 
-	local alternatives = alternative_units[unitDefID]
-	if not alternatives then
+	local pos = getCursorWorldPosition()
+	if not pos then
 		return
 	end
 
-	local mx, my = GetMouseState()
-	local _, coords = TraceScreenRay(mx, my, true, true)
-	if not coords then
+	local x, y, lmb, mmb, rmb = spGetMouseState()
+	local dist = distance2dSquared(mouseDownPos[1], mouseDownPos[3], pos[1], pos[3])
+
+	if mouseDownPos and lmb and dist > 100 then
+		-- currently doing a build drag, don't swap buildings
 		return
 	end
 
-	if TestBuildOrder(unitDefID, coords[1], coords[2], coords[3], 1) == 0 then
-		--Spring.Echo('cant build, looking for alternatives')
-		for _,alt_id in ipairs(alternatives) do --- try all alternatives
-			if TestBuildOrder(alt_id, coords[1], coords[2], coords[3], 1) ~= 0 then
-				if SetActiveCommand('buildunit_'..UnitDefs[alt_id].name) then
-					return
-				end
-			end
+	-- Check both arrays for a match with the current building cmd
+	local isGround = false
+	local alt = nil
+	local index = indexOf(groundBuildings, unitDefID)
```

Comment:
It's difficult to profile these things, but I believe it would be faster to get cursor world pos than iterating through a multiple arrays. Granted the arrays are both small, but that's generally a slow operation.